File: kernel/src/main/java/io/github/mosser/arduinoml/kernel/behavioral/UnaryExpression.java
package io.github.mosser.arduinoml.kernel.behavioral;

import io.github.mosser.arduinoml.kernel.generator.Visitable;
import io.github.mosser.arduinoml.kernel.generator.Visitor;
import io.github.mosser.arduinoml.kernel.structural.SIGNAL;
import io.github.mosser.arduinoml.kernel.structural.Sensor;

public class UnaryExpression implements Expression {

    private SIGNAL value;

    private Sensor sensor;

    public SIGNAL getValue() {
        return value;
    }

    public void setValue(SIGNAL value) {
        this.value = value;
    }

    public Sensor getSensor() {
        return sensor;
    }

    public void setSensor(Sensor sensor) {
        this.sensor = sensor;
    }

    @Override
    public void accept(Visitor visitor) {
        visitor.visit(this);
    }

    @Override
    public Expression getExpression() {
        return this;
    }

    @Override
    public String getCondition() {
        return "digitalRead(" + sensor.getPin() + ") == " + value;
    }
}

File: python/pyArduinoML/methodchaining/model/UnaryExpression.py
import sys
sys.path.append('..')
from model.Expression import Expression
from model.Sensor import Sensor
from model.SIGNAL import value

class UnaryExpression(Expression) :
    def __init__(self, sensor: Sensor, signal):
        self.sensor = sensor
        self.signal = signal
    
    def get_sensors(self):
        return [self.sensor]
    
    def __str__(self):
        return "digitalRead(%s) == %s" % (self.sensor, value(self.signal))
    
    def __repr__(self):
        return self.__str__()

    def __eq__(self, __o: object) -> bool:
        if isinstance(__o, UnaryExpression):
            return self.sensor == __o.sensor and self.signal == __o.signal